api/src/actions/custom-event.ts

```diff
@@ -2,23 +2,61 @@ import { dbConnect } from "@/src/utils/db-connect";
 import CustomEventModel from "@/src/models/custom-event";
 import { CustomEvent, EventEnvironment } from "@/src/utils/types";
 import CustomEventTypeModel from "@/src/models/custom-event-type";
-
+import { Types } from "mongoose";
+import { ObjectId } from 'mongodb'; // Import ObjectId from MongoDB driver
 
 export const createCustomEvent = async (event: Partial<CustomEvent>) => {
     await dbConnect();
-    let eventType = await CustomEventTypeModel.findOne({ _id: event.eventTypeId, projectId: event.projectId })
-
+    let eventId = new ObjectId(event.eventTypeId);
+    let eventType = await CustomEventTypeModel.findOne({ _id: eventId, projectId: event.projectId })
     if (!eventType) {
+        console.log("event type not found")
         return null;
     }
     let typeProperties = eventType.properties;
-    if (Object.keys(typeProperties).length === Object.keys((event?.properties as Record<string, string | number | Date>)).length
-        && Object.keys(typeProperties).every(k => event?.properties?.hasOwnProperty(k))) {
+    // if (Object.keys(typeProperties).length === Object.keys((event?.properties as Record<string, string | number | Date>)).length
+    //     && Object.keys(typeProperties).every(k => event?.properties?.hasOwnProperty(k))) {
+    //     return null;
+    // }
+    if (Object.keys(typeProperties).length !== Object.keys((event?.properties as Record<string, string | number | Date>)).length) {
+        console.log("properties length not equal")
         return null;
     }
+    for (const key in event?.properties) {
+        let has = false;
+        for (let i = 0; i < typeProperties.length; i++) {
+            let value = typeProperties[i]
+            console.log(value, value == key)
```

Comment:
personally - I would really be doing these checks in the controller so that you can throw relevant error messages there instead of returning null and throwing a generic error in the controller

Reply:
actions, IMO, are meant to be solely to do any relevant db queries and that's it - really shouldn't be doing additional checks here